File: 21.merge-two-sorted-lists.py
#
# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        # list nodeから配列に変える
        # list nodeに戻す
        if(not l1 and not l2):
            return None
        if(not l1 or not l2):
            return l1 or l2

        sorted_list = sorted(self.to_list(l1) + self.to_list(l2))
        return self.to_list_node(sorted_list)

    def to_list(self, node: ListNode):
        if(node.next is None):
            return [node.val]
        return [node.val] + self.to_list(node.next)

    def to_list_node(self, li):
        if(not li):
            return None
        return ListNode(li.pop(0), self.to_list_node(li))
# @lc code=end

File: solutions/21.merge-two-sorted-lists.ts
/*
 * @lc app=leetcode id=21 lang=typescript
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 */
//  class ListNode {
//      val: number
//      next: ListNode | null
//      constructor(val?: number, next?: ListNode | null) {
//          this.val = (val===undefined ? 0 : val)
//          this.next = (next===undefined ? null : next)
//      }
//  }

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  if(!list1) {
    return list2
  }
  if(!list2) {
    return list1
  }

  const toList = (node: ListNode): number[] => {
    if (!node.next) {
      return [node.val]
    }
    return [node.val, ...toList(node.next)]
  }

  const toNodeList = (list: number[]) => {
    if(!list) {
        return null
    }
    return new ListNode(list[0], toNodeList(list.slice(1)))
  }


  return toNodeList([...toList(list1), ...toList(list2)].sort())
};
// @lc code=end
